pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,449 @@
+            grads = self._single_shot_qnode_gradients(objective_fn, args, kwargs)
+        else:
+            raise ValueError(
+                "The objective function must either be encoded as a single QNode or "
+                "an ExpvalCost object for the Rosalin optimizer. "
+            )
+
+        # grads will have dimension [max(self.s), *params.shape]
+        # For each parameter, we want to truncate the number of shots to self.s[idx],
+        # and take the mean and the variance.
+        gradients = []
+        gradient_variances = []
+
+        for i, grad in enumerate(grads):
+            p_ind = np.ndindex(*grad.shape[1:])
+
+            g = np.zeros_like(grad[0])
+            s = np.zeros_like(grad[0])
+
+            for idx in p_ind:
+                grad_slice = grad[(slice(0, self.s[i][idx]),) + idx]
+                g[idx] = np.mean(grad_slice)
+                s[idx] = np.var(grad_slice, ddof=1)
+
+            gradients.append(g)
+            gradient_variances.append(s)
+
+        return gradients, gradient_variances
+
+    def step(self, objective_fn, *args, **kwargs):
+        """Update trainable arguments with one step of the optimizer.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            *args : Variable length argument list for objective function
+            **kwargs : variable length of keyword arguments for the objective function
+
+        Returns:
+            list[array]: the new variable values :math:`x^{(t+1)}`.
+            If single arg is provided, list [array] is replaced by array.
+        """
+
+        self.trainable_args = set()
+
+        for index, arg in enumerate(args):
+            if getattr(arg, "requires_grad", True):
+                self.trainable_args |= {index}
+
+        if self.s is None:
+            # Number of shots per parameter
+            self.s = [
+                np.zeros_like(a, dtype=np.int64) + self.min_shots
+                for i, a in enumerate(args)
+                if i in self.trainable_args
+            ]
+
+        # keep track of the number of shots run
+        s = np.concatenate([i.flatten() for i in self.s])
+        self.max_shots = max(s)
+        self.shots_used += int(2 * np.sum(s))
+
+        # compute the gradient, as well as the variance in the gradient,
+        # using the number of shots determined by the array s.
+        grads, grad_variances = self.compute_grad(objective_fn, args, kwargs)
+        new_args = self.apply_grad(grads, args)
+
+        if self.xi is None:
+            self.chi = [np.zeros_like(g, dtype=np.float64) for g in grads]
+            self.xi = [np.zeros_like(g, dtype=np.float64) for g in grads]
+
+        # running average of the gradient
+        self.chi = [self.mu * c + (1 - self.mu) * g for c, g in zip(self.chi, grads)]
+
+        # running average of the gradient variance
+        self.xi = [self.mu * x + (1 - self.mu) * v for x, v in zip(self.xi, grad_variances)]

```

Comment:
I am still not sure what "running" means...do you mean "in this step"?

Reply:
Oh, by running average I am meaning more 'moving average' or 'rolling average'. I think I used this term from the paper, but happy to change it to another one if it is less well known

Reply:
If it's from the paper that's fine I guess.